File: LeetCode/2/Solution.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbersCi(l1, l2, 0);
    }
    
    public ListNode addTwoNumbersCi(ListNode l1, ListNode l2, int ci) {
        ListNode head = null;
        if (l1 == null && l2 == null && ci == 0) {
            return head;
        } 
        l1 = l1 == null ? new ListNode(0) : l1;
        l2 = l2 == null ? new ListNode(0) : l2;
        head = new ListNode((l1.val + l2.val + ci) % 10);
        head.next = addTwoNumbersCi(l1.next, l2.next, (l1.val + l2.val + ci) / 10);
        return head;
    }
}

File: LeetCode/2/solution.py
class Solution:
    def addTwoNumbers(self, l1, l2):
        return self.addTwoNumbersCi(l1, l2, 0)

    def addTwoNumbersCi(self, l1, l2, ci):
        head = None
        if l1 == None and l2 == None and ci == 0:
            return head
        l1 = ListNode(0) if l1 == None else l1
        l2 = ListNode(0) if l2 == None else l2
        head = ListNode((l1.val + l2.val + ci) % 10)
        head.next = self.addTwoNumbersCi(l1.next, l2.next, int((l1.val + l2.val + ci) / 10))
        return head
